### ISSUE TITLE
Engines don't agree with each other on how to await promise instances with overridden `.then` properties

### ISSUE THREAD

**Issue by isiahmeadows** [tc39/ecma262#1577]
**Description:** Nobody agrees on how to handle overridden `.constructor`/`.then` properties on promise instances. The spec seems to side with V8 based on my interpretation of [`Await`](http://tc39.es/ecma262/#await), [`PromiseResolve`](http://tc39.es/ecma262/#sec-promise-resolve), and [promise resolve functions](http://tc39.es/ecma262/#sec-promise-resolve-functions), but [that breaks our ability to hook into promise chaining for scheduling implicit redraws](https://github.com/MithrilJS/mithril.js/issues/2426). Our implementation was written initially to work with userland promises, so it'd need updated to work in Chrome for async functions.

We *could* work around that with the current spec by also setting a `.constructor` property that delegates to the real constructor, basically `p.constructor = function (executor) { return new Promise(executor) }` any time we change `.then`. It was originally written prior to the async function proposal, and it could be easily modified.

[I initially filed a bug against V8](https://bugs.chromium.org/p/v8/issues/detail?id=9349), but after I read the spec, V8 appears to be in the right here, but it's breaking us in ways that make it look like a browser bug that has nothing to do with the underlying JS engine.

Test script:

```js
// test.js
const p = Promise.resolve()
const then = p.then

Object.defineProperty(p, "constructor", {
    get: () => {
        print("get constructor")
        return Promise
    }
})

p.then = (...args) => {
    print("received then")
    return Reflect.apply(then, p, args)
}

async function testAsync() {
    print("call async")
    await p
    print("resume await")
}

function testThen() {
    print("call then")
    return p.then(() => {
        print("resume then")
    })
}

testAsync()
.then(() => print("---------------"))
.then(testThen)
```

**eshost Output:**

```
$ eshost test.js
#### ch, xs
call async
get constructor
received then
get constructor
resume await
---------------
call then
received then
get constructor
resume then

#### jsc, sm
call async
received then
get constructor
resume await
---------------
call then
received then
get constructor
resume then

#### v8
call async
get constructor
resume await
---------------
call then
received then
get constructor
resume then
```

(Lightly edited the output for clarity on who aligns with who.)

**Comment by isiahmeadows** [comment]
I did find one genuine spec bug in the mix: `p.constructor === Promise` is *not* enough to ensure that `p` has all the fields of a promise (as [`PerformPromiseThen`](http://tc39.es/ecma262/#sec-performpromisethen) in step 9 of [the `Await` abstract operation](http://tc39.es/ecma262/#await) requires). That's all [`PromiseResolve` seems to check](http://tc39.es/ecma262/#sec-promise-resolve), and that's not sufficient when internal slots are in play.

**Comment by devsnek** [comment]
simplification of the above script which also shows the big concern:

```js
'use strict';

const p = Promise.resolve('x1');

Object.defineProperty(p, 'constructor', {
  get: () => {
    print('get constructor');
    return Promise;
  },
});

p.then = (onFulfilled) => {
  onFulfilled('x2');
  return Promise.resolve();
};

(async function () {
  const x = await p;
  return x;
}()).then(print);
```

```
┌────────────────────────┬─────────────────┐
│ Chakra                 │ get constructor │
│ XS                     │ x2              │
├────────────────────────┼─────────────────┤
│ engine262              │ get constructor │
│ V8                     │                 │
├────────────────────────┼─────────────────┤
│ JavaScriptCore         │ x2              │
│ SpiderMonkey           │                 │
└────────────────────────┴─────────────────┘
```

**Comment by mathiasbynens** [comment]
@isiahmeadows 

> (Lightly edited the output for clarity on who aligns with who.)

Pro tip: `eshost -s test.js` does this for you.

**Comment by mathiasbynens** [comment]
cc @domenic

**Comment by mathiasbynens** [comment]
TL;DR

- [`PromiseResolve(constructor, value)`](https://tc39.es/ecma262/#sec-promise-resolve) returns `value` if `constructor` is the original `Promise`, i.e. `%Promise%`.
- V8 and engine262 implement this per spec, but other engines have varying behavior.

I don’t think we’d want to change the spec here. Adding Test262 tests for this scenario (assuming they’re missing) would be a good next step and would help implementations align. @isiahmeadows would you be interested in turning your examples into a Test262 PR?

**Comment by isiahmeadows** [comment]
@mathiasbynens Sure.

**Comment by anba** [comment]
This difference is expected, see #1250.

The SpiderMonkey shell currently defaults to the previous state, but that'll change soon(ish): https://bugzilla.mozilla.org/show_bug.cgi?id=1558971
